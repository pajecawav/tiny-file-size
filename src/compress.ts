import { createReadStream } from "fs";
import { stat } from "fs/promises";
import { createBrotliCompress, createGzip } from "zlib";

export async function getPlainSize(file: string): Promise<number> {
	return (await stat(file)).size;
}

export function getGzipSize(file: string): Promise<number> {
	return new Promise((resolve, reject) => {
		let size = 0;

		createReadStream(file)
			.pipe(createGzip())
			.on("data", chunk => (size += chunk.length))
			.on("error", reject)
			.on("end", () => resolve(size));
	});
}

export function getBrotliSize(file: string): Promise<number> {
	return new Promise((resolve, reject) => {
		let size = 0;

		createReadStream(file)
			.pipe(createBrotliCompress())
			.on("data", chunk => (size += chunk.length))
			.on("error", reject)
			.on("end", () => resolve(size));
	});
}
