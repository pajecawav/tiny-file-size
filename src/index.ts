export { getBrotliSize, getGzipSize, getRawSize } from "./compress";
export { buildJsonReport, buildPrettyReport, humanizeBytes } from "./report";
export type { PrettyReportOptions } from "./report";
export { getFileSizes, run } from "./run";
export type { FileSize, FileSizesOptions } from "./run";
