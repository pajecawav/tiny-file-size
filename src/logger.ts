import { bgRed, black, bold, red } from "picocolors";

export const logger = {
	error(message: string) {
		console.error(bgRed(bold(black(" ERROR "))), red(message));
	},
};
