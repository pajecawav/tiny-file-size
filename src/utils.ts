import binaryExtensions from "binary-extensions";
import { extname } from "path";

const binaryExtensionsSet = new Set(binaryExtensions);

export function sum(values: number[]): number {
	let total = 0;
	for (const value of values) total += value;
	return total;
}

export function isBinaryFile(path: string): boolean {
	return binaryExtensionsSet.has(extname(path).slice(1));
}
