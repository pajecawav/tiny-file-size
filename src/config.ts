import colors from "picocolors";
import { version } from "../package.json";
import { logger } from "./logger";

export interface Config {
	globs: string[];
	gzip: boolean;
	brotli: boolean;
	output: "json" | "pretty";
	total: boolean;
}

const HELP = `
${colors.bold("Usage:")}
  file-size [options] [globs...]

${colors.bold("Options:")}
  -h, --help                     Show help
  -g, --gzip                     Include gzip size in output 
  -b, --brotli                   Include brotli size in output 
      --total                    Include total size in output
      --json                     Print result as json
  -v, --version                  Print the current version and exit

${colors.bold("Examples:")}
  # Show sizes of files foo.js and bar.json
  file-size foo.js bar.json

  # Include sizes after gzip and brotli
  file-size --gzip --brotli foo.js bar.json

  # Print output as JSON
  file-size --json foo.js bar.json
`.trim();

export function parseArgs(args: string[]): Config {
	const cfg: Config = {
		globs: [],
		gzip: false,
		brotli: false,
		output: "pretty",
		total: false,
	};

	for (let i = 0; i < args.length; i++) {
		const arg = args[i];

		if (arg[0] !== "-") {
			cfg.globs = args.slice(i);
			break;
		}

		switch (arg) {
			case "-h":
			case "--help":
				console.log(HELP);
				process.exit(0);
				break;
			case "-v":
			case "--version":
				console.log(version);
				process.exit(0);
				break;
			case "-g":
			case "--gzip":
				cfg.gzip = true;
				break;
			case "-b":
			case "--brotli":
				cfg.brotli = true;
				break;
			case "--json":
				cfg.output = "json";
				break;
			case "--total":
				cfg.total = true;
				break;
			default:
				logger.error(`Unknown argument ${arg}`);
				process.exit(1);
		}
	}

	return cfg;
}

export const config = parseArgs(process.argv.slice(2));
