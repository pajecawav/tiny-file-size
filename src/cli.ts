import { config } from "./config";
import { run } from "./run";

run(config);
