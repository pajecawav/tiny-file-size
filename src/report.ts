import path from "path";
import { bold, dim, green } from "picocolors";
import { FileSize } from "./run";

export function reportJson(sizes: FileSize[]) {
	console.log(sizes);
}

export function reportConsole(sizes: FileSize[]) {
	for (const { file, plain, gzip, brotli } of sizes) {
		const { dir, base } = path.parse(file);

		const pathStr = (dir ? dim(dir + "/") : "") + green(base);

		const plainStr = humanizeBytes(plain);
		const gzipStr = humanizeBytes(gzip);
		const brotliStr = humanizeBytes(brotli);

		console.log(`${pathStr}: ${bold(plainStr)} (${gzipStr} gzip, ${brotliStr} brotli)`);
	}
}

function humanizeBytes(bytes: number): string {
	let suffix = "MB";
	for (const suff of ["B", "KB"]) {
		if (bytes < 1024) {
			suffix = suff;
			break;
		}
		bytes /= 1024;
	}

	if (suffix === "B") {
		return `${bytes.toFixed(0)} ${suffix}`;
	}

	return `${bytes.toFixed(2)} ${suffix}`;
}
