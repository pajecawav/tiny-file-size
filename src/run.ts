import { existsSync, statSync } from "fs";
import { getBrotliSize, getGzipSize, getPlainSize } from "./compress";
import { Config } from "./config";
import { logger } from "./logger";
import { buildPrettyReport, buildJsonReport } from "./report";

export interface FileSize {
	file: string;
	plain: number;
	gzip: number | null;
	brotli: number | null;
}

interface FileSizesOptions {
	files: string[];
	gzip: boolean;
	brotli: boolean;
}

export async function getFileSizes(config: FileSizesOptions): Promise<FileSize[]> {
	const sizes: FileSize[] = await Promise.all(
		config.files.map(async file => {
			const [plain, gzip, brotli] = await Promise.all([
				getPlainSize(file),
				config.gzip ? getGzipSize(file) : null,
				config.brotli ? getBrotliSize(file) : null,
			]);

			return { file, plain, gzip, brotli };
		})
	);

	return sizes;
}

export async function run(config: Config) {
	for (const file of config.files) {
		if (!existsSync(file) || !statSync(file).isFile()) {
			logger.error(`No such file ${file}`);
			process.exit(1);
		}
	}

	let sizes: FileSize[];
	try {
		sizes = await getFileSizes(config);
	} catch (e: unknown) {
		if (!(e instanceof Error)) {
			throw e;
		}

		logger.error(e.message);
		process.exit(1);
	}

	if (config.output === "json") {
		console.log(buildJsonReport(sizes));
	} else {
		console.log(buildPrettyReport(sizes));
	}
}
