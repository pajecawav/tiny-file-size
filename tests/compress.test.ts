import { describe, expect, it } from "vitest";
import { getBrotliSize, getGzipSize, getPlainSize } from "../src/compress";
import { files } from "./fixtures/files";

describe("compress", () => {
	it("calculates plain size", async () => {
		const sizes = await Promise.all(files.map(file => getPlainSize(file)));
		expect(sizes).toMatchSnapshot();
	});

	it("compresses with gzip", async () => {
		const sizes = await Promise.all(files.map(file => getGzipSize(file)));
		expect(sizes).toMatchSnapshot();
	});

	it("compresses with brotli", async () => {
		const sizes = await Promise.all(files.map(file => getBrotliSize(file)));
		expect(sizes).toMatchSnapshot();
	});
});
