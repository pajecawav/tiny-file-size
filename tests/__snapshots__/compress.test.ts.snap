// Vitest Snapshot v1

exports[`compress > calculates plain size 1`] = `
[
  2519,
  4183,
  2877,
]
`;

exports[`compress > compresses with brotli 1`] = `
[
  1395,
  2193,
  1142,
]
`;

exports[`compress > compresses with gzip 1`] = `
[
  1484,
  2410,
  1205,
]
`;
